{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Purpose:\n",
    "- To add DLC eye data to design matrix\n",
    "- As in bod.eye_tracking\n",
    "    - Should have at least pupil_area, eye_width, and likely_blink\n",
    "- See if implementing into COMB is easy (patch)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Conclusion:\n",
    "- package issue, involving pytables\n",
    "    - pd.read_hdf does not work or errors out\n",
    "- Hold this off for now, continue dev GLM fitting"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import h5py\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from pathlib import Path\n",
    "\n",
    "# notebook dev\n",
    "%load_ext autoreload\n",
    "%autoreload 2\n",
    "%matplotlib inline\n",
    "import capsule_utils\n",
    "import load_data\n",
    "import design_matrix_tools as dmtools\n",
    "import kernel_tools as ktools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File rig.json not found in /root/capsule/data/multiplane-ophys_721291_2024-05-16_08-57-00\n"
     ]
    }
   ],
   "source": [
    "session_name = 'multiplane-ophys_721291_2024-05-16_08-57-00'\n",
    "\n",
    "data_dir = Path('/root/capsule/data/')\n",
    "raw_path = data_dir / session_name\n",
    "processed_path = list(data_dir.glob(f'{session_name}_processed*'))[0]\n",
    "\n",
    "opids = []\n",
    "for plane_folder in processed_path.glob(\"*\"):\n",
    "    if plane_folder.is_dir() and not plane_folder.stem.startswith(\"nextflow\"):\n",
    "        opid = plane_folder.stem\n",
    "        opids.append(opid)\n",
    "\n",
    "# bod_list = []\n",
    "# for opid in opids:\n",
    "#     bod = load_data.load_plane_data(session_name, opid=opid)\n",
    "#     bod = capsule_utils.add_trials_to_bod(bod)\n",
    "#     bod_list.append(bod)\n",
    "# bod = bod_list[0]\n",
    "bod = load_data.load_plane_data(session_name, opid=opids[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<KeysViewHDF5 ['cr', 'eye', 'pupil']>\n",
      "<KeysViewHDF5 ['axis0', 'axis1', 'block0_items', 'block0_values', 'block1_items', 'block1_values']>\n",
      "<KeysViewHDF5 ['axis0', 'axis1', 'block0_items', 'block0_values', 'block1_items', 'block1_values']>\n",
      "[b'center_x' b'center_y' b'width' b'height' b'phi'\n",
      " b'pupil_average_confidence' b'pupil_area' b'pupil_is_bad_frame']\n"
     ]
    }
   ],
   "source": [
    "eye_dir = list(data_dir.glob(f'{session_name}_dlc-eye_*'))\n",
    "assert len(eye_dir) == 1\n",
    "\n",
    "eye_h5_fn = eye_dir[0] / 'ellipses_processed.h5'\n",
    "# eye_h5_fn = eye_dir[0] / 'ellipses.h5'\n",
    "with h5py.File(eye_h5_fn, 'r') as h:\n",
    "    print(h.keys())\n",
    "    print(h['pupil'].keys())\n",
    "    print(h['eye'].keys())\n",
    "    print(h['pupil']['axis0'][:])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<KeysViewHDF5 ['df_with_missing']>\n",
      "<KeysViewHDF5 ['_i_table', 'table']>\n",
      "[(     0, [ 6.33884335e+00,  7.72512207e+01,  1.82653416e-03,  6.24676666e+01,  1.99652130e+02,  3.65068251e-03,  5.97942200e+01,  2.05098557e+02,  5.09457104e-03,  6.19379807e+01,  2.00198074e+02,  2.04754993e-02,  6.09929810e+01,  1.97884903e+02,  1.46581922e-02,  6.47720032e+01,  1.73709076e+02,  1.47243375e-02,  6.33734016e+01,  1.65902313e+02,  1.24610318e-02,  5.77698784e+01,  1.92075073e+02,  6.78498857e-03,  5.87980461e+01,  1.92362122e+02,  1.86867197e-03,  3.26299095e+00,  1.55936676e+02,  1.23810431e-03,  5.22358751e+00,  7.45234756e+01,  7.52468535e-04,  6.16038208e+01,  1.96104324e+02,  1.27416616e-03, -3.07590961e-01,  1.53588181e+02,  4.12408775e-03,  4.73304987e+00,  1.56218582e+02,  1.98574411e-03,  3.57965565e+00,  3.33953953e+00,  2.16960185e-03,  9.56604004e+01,  2.04954910e+02,  5.03960298e-03,  4.57403564e+00,  8.73029938e+01,  2.37481464e-02,  4.22407799e+01,  2.02470993e+02,  6.65119290e-03,  4.79362297e+00,  1.57717850e+02,  3.99512723e-02,  5.10478210e+01,  1.96744720e+02,  1.33928740e-02,  5.19553614e+00,  1.53566910e+02,  4.43067364e-02,  4.63762712e+00,  1.54431839e+02,  7.61897489e-03,  2.51432610e+00,  1.59354614e+02,  9.24489740e-03,  6.05332375e+00,  1.90352829e+02,  1.55064405e-03,  4.48864174e+00,  7.82162704e+01,  6.43228274e-03,  2.46735811e+00,  7.83396912e+01,  6.67607132e-03,  5.23082008e+01,  1.88080811e+02,  4.02670214e-03,  6.53436661e+01,  1.88825439e+02,  7.95483775e-03,  6.29269981e+01,  1.92569473e+02,  3.55931669e-02,  6.20461273e+01,  1.94868500e+02,  3.70094948e-03,  5.88337374e+00,  1.90140030e+02,  2.15125829e-03,  5.73908615e+00,  1.82166199e+02,  2.98997178e-03,  4.06040287e+00,  1.55930573e+02,  9.23890714e-03,  6.27971306e+01,  1.90228577e+02,  2.14701984e-03,  8.62267971e-01,  7.37446823e+01,  1.96577446e-03,  3.26474953e+00,  7.52736206e+01,  4.56767483e-03])\n",
      " (     1, [ 3.43134796e+02,  1.89704971e+02,  3.44079621e-02,  3.41375397e+02,  1.89134964e+02,  6.76303059e-02,  3.40345825e+02,  1.91940659e+02,  5.79203628e-02,  3.37495514e+02,  1.90912613e+02,  6.04924001e-02,  3.32902527e+02,  1.92153107e+02,  5.97173236e-02,  3.33641937e+02,  1.89027115e+02,  5.90337031e-02,  3.34733398e+02,  1.88242752e+02,  3.36096473e-02,  3.35334137e+02,  1.83634262e+02,  2.05148011e-02,  3.34932526e+02,  1.82225281e+02,  1.84112322e-02,  3.38192963e+02,  1.83218048e+02,  1.53782452e-02,  3.37869049e+02,  1.80552124e+02,  1.46369534e-02,  3.46077942e+02,  1.91198486e+02,  2.27050725e-02,  4.46050049e+02,  2.26013702e+02,  2.00575531e-01,  4.24315918e+02,  2.62491730e+02,  6.54620588e-01,  3.80344269e+02,  2.90601318e+02,  9.52759147e-01,  3.19551117e+02,  2.91205353e+02,  8.06727767e-01,  3.75194092e+02,  2.94784393e+02,  1.07953325e-01,  1.91458054e+02,  1.82531708e+02,  2.37825021e-01,  1.81858643e+02,  1.38597946e+02,  4.49471176e-01,  2.38533203e+02,  1.06962822e+02,  1.65838018e-01,  2.82938568e+02,  8.39939117e+01,  3.86750922e-02,  2.05713837e+02,  7.91414337e+01,  3.87314819e-02,  2.51220398e+02,  6.51652222e+01,  5.97943328e-02,  3.30891815e+02,  1.69275085e+02,  9.87290055e-04,  3.56134277e+02,  1.88214630e+02,  2.86504766e-03,  3.44518005e+02,  1.98833649e+02,  5.19835856e-03,  3.45005890e+02,  1.92761948e+02,  2.50138156e-03,  3.32259064e+02,  1.85460144e+02,  1.45709433e-03,  6.55661316e+02,  1.21248657e+02,  2.96694203e-03,  3.31018066e+02,  1.85331268e+02,  9.61109006e-04,  3.27261688e+02,  1.74473495e+02,  9.06262139e-04,  3.27983673e+02,  1.72684586e+02,  1.80772971e-03,  3.30864502e+02,  1.67058594e+02,  8.28336086e-03,  3.33949951e+02,  1.71046646e+02,  6.20628754e-03,  3.49854767e+02,  1.67615524e+02,  3.15150525e-03,  3.50168488e+02,  1.72608627e+02,  3.95032298e-03])\n",
      " (     2, [ 3.43757538e+02,  1.89520874e+02,  2.12395694e-02,  3.42846802e+02,  1.88985580e+02,  4.18222547e-02,  3.41265472e+02,  1.91695999e+02,  3.94434221e-02,  3.39257996e+02,  1.91788940e+02,  2.73177437e-02,  3.35542999e+02,  1.90731796e+02,  2.98347417e-02,  3.35078217e+02,  1.88745453e+02,  2.48713810e-02,  3.34945770e+02,  1.87694931e+02,  1.64309815e-02,  3.36843506e+02,  1.86622162e+02,  7.54291005e-03,  3.38401337e+02,  1.83372681e+02,  7.26259872e-03,  3.39507782e+02,  1.78634308e+02,  7.20435986e-03,  3.44943939e+02,  1.80366013e+02,  7.11477641e-03,  3.45819855e+02,  1.89720505e+02,  1.45007931e-02,  4.63301361e+02,  1.96753036e+02,  4.02013570e-01,  4.36833649e+02,  2.54038620e+02,  6.38396323e-01,  3.86903320e+02,  2.88947174e+02,  7.30720878e-01,  3.39251556e+02,  2.95994873e+02,  3.06748420e-01,  2.78741882e+02,  2.80579224e+02,  8.09696093e-02,  3.37852997e+02,  2.50500443e+02,  1.59644738e-01,  1.82103287e+02,  1.40991486e+02,  5.38076699e-01,  2.54251801e+02,  1.06944984e+02,  1.03556283e-01,  2.97371582e+02,  9.72081680e+01,  2.28320230e-02,  2.11161621e+02,  8.10720596e+01,  2.39137374e-02,  2.51882721e+02,  6.49492493e+01,  2.34728474e-02,  4.36993134e+02,  9.41728745e+01,  5.19413874e-03,  3.63825989e+02,  1.85785873e+02,  5.11191972e-03,  3.45652344e+02,  1.99249252e+02,  4.80828295e-03,  3.45189667e+02,  1.93087250e+02,  1.73635967e-03,  3.44489166e+02,  1.96038330e+02,  8.14423722e-04,  6.53527649e+02,  1.32213318e+02,  9.51258698e-04,  3.42973724e+02,  1.96840515e+02,  4.39802883e-04,  3.26361725e+02,  1.75800659e+02,  7.27835635e-04,  3.26533478e+02,  1.73593765e+02,  1.52326690e-03,  3.29197723e+02,  1.68096100e+02,  6.76685618e-03,  3.33353607e+02,  1.70489059e+02,  5.83607238e-03,  3.49185883e+02,  1.67380661e+02,  3.59999225e-03,  3.57234863e+02,  1.71202866e+02,  4.20754077e-03])\n",
      " ...\n",
      " (271836, [ 3.40835754e+02,  1.91620483e+02,  9.96539950e-01,  3.39552704e+02,  1.95745834e+02,  9.92453277e-01,  3.37096649e+02,  1.97924286e+02,  9.91839111e-01,  3.32678619e+02,  1.98606949e+02,  9.92565870e-01,  3.27275177e+02,  1.97351715e+02,  9.90230143e-01,  3.24942993e+02,  1.93781403e+02,  9.92377281e-01,  3.24222778e+02,  1.91100830e+02,  9.95416641e-01,  3.25814331e+02,  1.85478958e+02,  9.95512664e-01,  3.29760284e+02,  1.83316772e+02,  9.97181773e-01,  3.34176514e+02,  1.83003891e+02,  9.97228444e-01,  3.38201843e+02,  1.83915436e+02,  9.95241284e-01,  3.40170197e+02,  1.86974060e+02,  9.94809031e-01,  4.71572021e+02,  1.77052200e+02,  9.99931574e-01,  4.51512543e+02,  2.34591736e+02,  9.99902129e-01,  4.01017609e+02,  2.77301453e+02,  9.99920964e-01,  3.23373993e+02,  2.92062103e+02,  9.99997020e-01,  2.51236725e+02,  2.78362701e+02,  9.99994874e-01,  1.97057114e+02,  2.35791458e+02,  9.99996662e-01,  1.77899338e+02,  1.81641754e+02,  9.99915123e-01,  2.01763107e+02,  1.25060135e+02,  9.99970555e-01,  2.64547729e+02,  8.62570190e+01,  9.99963164e-01,  3.39762329e+02,  7.84095993e+01,  9.99985933e-01,  4.06498444e+02,  9.40865479e+01,  9.99988437e-01,  4.51502380e+02,  1.25741356e+02,  9.99956846e-01,  3.67776642e+02,  1.65299759e+02,  9.99620318e-01,  3.63067627e+02,  1.81691849e+02,  9.98753786e-01,  3.48610840e+02,  1.93969025e+02,  9.95682120e-01,  3.32031372e+02,  1.99657486e+02,  9.98582482e-01,  3.10887695e+02,  1.96997528e+02,  9.96116161e-01,  2.95433044e+02,  1.85507172e+02,  9.97033954e-01,  2.90162537e+02,  1.66219482e+02,  9.95735645e-01,  2.94974121e+02,  1.46825134e+02,  9.95333970e-01,  3.12370667e+02,  1.31567917e+02,  9.97705519e-01,  3.31726135e+02,  1.26919594e+02,  9.99385715e-01,  3.52798737e+02,  1.35321335e+02,  9.98852730e-01,  3.64684998e+02,  1.49071884e+02,  9.99610603e-01])\n",
      " (271837, [ 3.41433319e+02,  1.91235550e+02,  9.96370554e-01,  3.40128876e+02,  1.95068756e+02,  9.91629779e-01,  3.37226776e+02,  1.97769180e+02,  9.92369413e-01,  3.32958099e+02,  1.98352753e+02,  9.92315888e-01,  3.27727905e+02,  1.97298965e+02,  9.88674223e-01,  3.25565582e+02,  1.93760742e+02,  9.90639448e-01,  3.24311401e+02,  1.90022446e+02,  9.93611038e-01,  3.26246735e+02,  1.85635513e+02,  9.94456887e-01,  3.30252594e+02,  1.83606003e+02,  9.96903598e-01,  3.34669006e+02,  1.83280960e+02,  9.96586084e-01,  3.38667542e+02,  1.84088425e+02,  9.94311690e-01,  3.40770691e+02,  1.86721420e+02,  9.94423807e-01,  4.72090149e+02,  1.76222717e+02,  9.99967933e-01,  4.52838928e+02,  2.33565918e+02,  9.99878407e-01,  4.01553223e+02,  2.77098450e+02,  9.99945760e-01,  3.24877014e+02,  2.91550812e+02,  9.99996305e-01,  2.51274002e+02,  2.77435760e+02,  9.99995828e-01,  1.98031021e+02,  2.35477966e+02,  9.99996424e-01,  1.78160049e+02,  1.81262482e+02,  9.99927878e-01,  2.02572174e+02,  1.24183273e+02,  9.99957681e-01,  2.65159363e+02,  8.73490295e+01,  9.99969363e-01,  3.40433411e+02,  7.82430420e+01,  9.99968648e-01,  4.07395081e+02,  9.50419312e+01,  9.99975920e-01,  4.51994843e+02,  1.25547714e+02,  9.99971390e-01,  3.68805603e+02,  1.63578461e+02,  9.99543607e-01,  3.63697998e+02,  1.81203339e+02,  9.98130143e-01,  3.51608093e+02,  1.92594833e+02,  9.95361626e-01,  3.34856140e+02,  1.99806519e+02,  9.98875678e-01,  3.13719482e+02,  1.97146912e+02,  9.97331858e-01,  2.99310120e+02,  1.86471085e+02,  9.97767448e-01,  2.92086792e+02,  1.69162262e+02,  9.97508168e-01,  2.95104553e+02,  1.49612167e+02,  9.97507811e-01,  3.11307983e+02,  1.32577469e+02,  9.97876287e-01,  3.30917725e+02,  1.27265335e+02,  9.99473512e-01,  3.52016693e+02,  1.34406296e+02,  9.99021053e-01,  3.64819824e+02,  1.47426956e+02,  9.99620438e-01])\n",
      " (271838, [ 3.41284668e+02,  1.91187164e+02,  9.95800793e-01,  3.39924011e+02,  1.94940125e+02,  9.90410805e-01,  3.36803284e+02,  1.97498413e+02,  9.92616832e-01,  3.32751892e+02,  1.97944412e+02,  9.93055165e-01,  3.27866577e+02,  1.96963638e+02,  9.90041673e-01,  3.25773712e+02,  1.93380463e+02,  9.91636813e-01,  3.24778595e+02,  1.89872940e+02,  9.94136810e-01,  3.26712982e+02,  1.85481552e+02,  9.94651914e-01,  3.30460083e+02,  1.83550171e+02,  9.96150494e-01,  3.34897003e+02,  1.83266251e+02,  9.95591521e-01,  3.38702240e+02,  1.84426224e+02,  9.92536306e-01,  3.40756958e+02,  1.86693298e+02,  9.93262708e-01,  4.73292450e+02,  1.78236191e+02,  9.99920964e-01,  4.52262695e+02,  2.34891891e+02,  9.99899864e-01,  4.00889374e+02,  2.78160370e+02,  9.99935150e-01,  3.23132507e+02,  2.91661774e+02,  9.99996066e-01,  2.50867813e+02,  2.77103058e+02,  9.99995351e-01,  1.98919601e+02,  2.36761856e+02,  9.99996901e-01,  1.78452347e+02,  1.82129425e+02,  9.99893308e-01,  2.01937271e+02,  1.24998108e+02,  9.99951124e-01,  2.64919403e+02,  8.78621445e+01,  9.99961376e-01,  3.41231873e+02,  7.87508621e+01,  9.99976993e-01,  4.08997742e+02,  9.51645355e+01,  9.99987364e-01,  4.53410583e+02,  1.25737892e+02,  9.99980330e-01,  3.68885834e+02,  1.62625519e+02,  9.99594629e-01,  3.65204803e+02,  1.80102081e+02,  9.98545051e-01,  3.51304016e+02,  1.92858231e+02,  9.96299207e-01,  3.35990753e+02,  1.99533356e+02,  9.98488665e-01,  3.14375427e+02,  1.97784927e+02,  9.97202039e-01,  3.00047821e+02,  1.87154861e+02,  9.97844100e-01,  2.92315430e+02,  1.69821777e+02,  9.96878862e-01,  2.95473541e+02,  1.50307617e+02,  9.97358739e-01,  3.10681763e+02,  1.33472992e+02,  9.98191774e-01,  3.29987305e+02,  1.27992828e+02,  9.99497652e-01,  3.51918091e+02,  1.34405075e+02,  9.99159813e-01,  3.64317719e+02,  1.46858231e+02,  9.99744236e-01])]\n"
     ]
    }
   ],
   "source": [
    "eye_raw_fn = list(eye_dir[0].glob('*resnet50_*.h5'))[0]\n",
    "with h5py.File(eye_raw_fn, 'r') as h:\n",
    "    print(h.keys())\n",
    "    print(h['df_with_missing'].keys())\n",
    "    print(h['df_with_missing']['table'][:])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "def load_eye_tracking_hdf(eye_tracking_file: Path) -> pd.DataFrame:\n",
    "    \"\"\"Load a DeepLabCut hdf5 file containing eye tracking data into a\n",
    "    dataframe.\n",
    "\n",
    "    Note: The eye tracking hdf5 file contains 3 separate dataframes. One for\n",
    "    corneal reflection (cr), eye, and pupil ellipse fits. This function\n",
    "    loads and returns this data as a single dataframe.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    eye_tracking_file : Path\n",
    "        Path to an hdf5 file produced by the DeepLabCut eye tracking pipeline.\n",
    "        The hdf5 file will contain the following keys: \"cr\", \"eye\", \"pupil\".\n",
    "        Each key has an associated dataframe with the following\n",
    "        columns: \"center_x\", \"center_y\", \"height\", \"width\", \"phi\".\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    pd.DataFrame\n",
    "        A dataframe containing combined corneal reflection (cr), eyelid (eye),\n",
    "        and pupil data. Column names for each field will be renamed by\n",
    "        prepending the field name. (e.g. center_x -> eye_center_x)\n",
    "    \"\"\"\n",
    "    eye_tracking_fields = [\"cr\", \"eye\", \"pupil\"]\n",
    "\n",
    "    eye_tracking_dfs = []\n",
    "    for field_name in eye_tracking_fields:\n",
    "        field_data = pd.read_hdf(eye_tracking_file, key=field_name)\n",
    "        new_col_name_map = {col_name: f\"{field_name}_{col_name}\"\n",
    "                            for col_name in field_data.columns}\n",
    "        field_data.rename(new_col_name_map, axis=1, inplace=True)\n",
    "        eye_tracking_dfs.append(field_data)\n",
    "\n",
    "    eye_tracking_data = pd.concat(eye_tracking_dfs, axis=1)\n",
    "    eye_tracking_data.index.name = 'frame'\n",
    "\n",
    "    # Values in the hdf5 may be complex (likely an artifact of the ellipse\n",
    "    # fitting process). Take only the real component.\n",
    "    eye_tracking_data = eye_tracking_data.apply(lambda x: np.real(x.to_numpy()))  # noqa: E501\n",
    "\n",
    "    return eye_tracking_data.astype(float)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[autoreload of _frozen_importlib failed: Traceback (most recent call last):\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 273, in check\n",
      "    superreload(m, reload, self.old_objects)\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 471, in superreload\n",
      "    module = reload(module)\n",
      "  File \"/opt/conda/lib/python3.9/importlib/__init__.py\", line 166, in reload\n",
      "    spec = module.__spec__ = _bootstrap._find_spec(name, pkgpath, target)\n",
      "AttributeError: module 'importlib._bootstrap' has no attribute '_find_spec'\n",
      "]\n",
      "[autoreload of _frozen_importlib_external failed: Traceback (most recent call last):\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 273, in check\n",
      "    superreload(m, reload, self.old_objects)\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 471, in superreload\n",
      "    module = reload(module)\n",
      "  File \"/opt/conda/lib/python3.9/importlib/__init__.py\", line 166, in reload\n",
      "    spec = module.__spec__ = _bootstrap._find_spec(name, pkgpath, target)\n",
      "  File \"<frozen importlib._bootstrap>\", line 925, in _find_spec\n",
      "  File \"<frozen importlib._bootstrap_external>\", line 1423, in find_spec\n",
      "  File \"<frozen importlib._bootstrap_external>\", line 1390, in _get_spec\n",
      "NameError: name 'isinstance' is not defined\n",
      "]\n",
      "[autoreload of codecs failed: Traceback (most recent call last):\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 273, in check\n",
      "    superreload(m, reload, self.old_objects)\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 496, in superreload\n",
      "    update_generic(old_obj, new_obj)\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 393, in update_generic\n",
      "    update(a, b)\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 361, in update_class\n",
      "    update_instances(old, new)\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 319, in update_instances\n",
      "    object.__setattr__(ref, \"__class__\", new)\n",
      "TypeError: __class__ assignment: 'CodecInfo' object layout differs from 'CodecInfo'\n",
      "]\n",
      "[autoreload of abc failed: Traceback (most recent call last):\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 273, in check\n",
      "    superreload(m, reload, self.old_objects)\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 496, in superreload\n",
      "    update_generic(old_obj, new_obj)\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 393, in update_generic\n",
      "    update(a, b)\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 361, in update_class\n",
      "    update_instances(old, new)\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 319, in update_instances\n",
      "    object.__setattr__(ref, \"__class__\", new)\n",
      "TypeError: can't apply this __setattr__ to ABCMeta object\n",
      "]\n",
      "[autoreload of _collections_abc failed: Traceback (most recent call last):\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 273, in check\n",
      "    superreload(m, reload, self.old_objects)\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 471, in superreload\n",
      "    module = reload(module)\n",
      "  File \"/opt/conda/lib/python3.9/importlib/__init__.py\", line 148, in reload\n",
      "    raise ImportError(msg.format(name), name=name)\n",
      "ImportError: module collections.abc not in sys.modules\n",
      "]\n",
      "[autoreload of types failed: Traceback (most recent call last):\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 273, in check\n",
      "    superreload(m, reload, self.old_objects)\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 471, in superreload\n",
      "    module = reload(module)\n",
      "  File \"/opt/conda/lib/python3.9/importlib/__init__.py\", line 139, in reload\n",
      "    if not module or not isinstance(module, types.ModuleType):\n",
      "AttributeError: module 'types' has no attribute 'ModuleType'\n",
      "]\n",
      "[autoreload of importlib failed: Traceback (most recent call last):\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 273, in check\n",
      "    superreload(m, reload, self.old_objects)\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 471, in superreload\n",
      "    module = reload(module)\n",
      "  File \"/opt/conda/lib/python3.9/importlib/__init__.py\", line 139, in reload\n",
      "    if not module or not isinstance(module, types.ModuleType):\n",
      "NameError: name 'isinstance' is not defined\n",
      "]\n",
      "[autoreload of enum failed: Traceback (most recent call last):\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 273, in check\n",
      "    superreload(m, reload, self.old_objects)\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 496, in superreload\n",
      "    update_generic(old_obj, new_obj)\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 393, in update_generic\n",
      "    update(a, b)\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 326, in update_class\n",
      "    old_obj = getattr(old, key)\n",
      "  File \"/opt/conda/lib/python3.9/enum.py\", line 429, in __getattr__\n",
      "    raise AttributeError(name) from None\n",
      "AttributeError: name\n",
      "]\n",
      "[autoreload of sre_constants failed: Traceback (most recent call last):\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 273, in check\n",
      "    superreload(m, reload, self.old_objects)\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 496, in superreload\n",
      "    update_generic(old_obj, new_obj)\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 393, in update_generic\n",
      "    update(a, b)\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 361, in update_class\n",
      "    update_instances(old, new)\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 319, in update_instances\n",
      "    object.__setattr__(ref, \"__class__\", new)\n",
      "TypeError: __class__ assignment: '_NamedIntConstant' object layout differs from '_NamedIntConstant'\n",
      "]\n",
      "[autoreload of re failed: Traceback (most recent call last):\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 273, in check\n",
      "    superreload(m, reload, self.old_objects)\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 496, in superreload\n",
      "    update_generic(old_obj, new_obj)\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 393, in update_generic\n",
      "    update(a, b)\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 361, in update_class\n",
      "    update_instances(old, new)\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 319, in update_instances\n",
      "    object.__setattr__(ref, \"__class__\", new)\n",
      "TypeError: __class__ assignment: 'RegexFlag' object layout differs from 'RegexFlag'\n",
      "]\n",
      "[autoreload of typing failed: Traceback (most recent call last):\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 273, in check\n",
      "    superreload(m, reload, self.old_objects)\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 471, in superreload\n",
      "    module = reload(module)\n",
      "  File \"/opt/conda/lib/python3.9/importlib/__init__.py\", line 169, in reload\n",
      "    _bootstrap._exec(spec, module)\n",
      "  File \"<frozen importlib._bootstrap>\", line 613, in _exec\n",
      "  File \"<frozen importlib._bootstrap_external>\", line 850, in exec_module\n",
      "  File \"<frozen importlib._bootstrap>\", line 228, in _call_with_frames_removed\n",
      "  File \"/opt/conda/lib/python3.9/typing.py\", line 1692, in <module>\n",
      "    Hashable = _alias(collections.abc.Hashable, 0)  # Not generic.\n",
      "  File \"/opt/conda/lib/python3.9/collections/__init__.py\", line 68, in __getattr__\n",
      "    raise AttributeError(f'module {__name__!r} has no attribute {name!r}')\n",
      "AttributeError: module 'collections' has no attribute 'abc'\n",
      "]\n",
      "[autoreload of signal failed: Traceback (most recent call last):\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 273, in check\n",
      "    superreload(m, reload, self.old_objects)\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 496, in superreload\n",
      "    update_generic(old_obj, new_obj)\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 393, in update_generic\n",
      "    update(a, b)\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 361, in update_class\n",
      "    update_instances(old, new)\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 319, in update_instances\n",
      "    object.__setattr__(ref, \"__class__\", new)\n",
      "TypeError: __class__ assignment: 'Signals' object layout differs from 'Signals'\n",
      "]\n",
      "[autoreload of platform failed: Traceback (most recent call last):\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 273, in check\n",
      "    superreload(m, reload, self.old_objects)\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 496, in superreload\n",
      "    update_generic(old_obj, new_obj)\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 393, in update_generic\n",
      "    update(a, b)\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 361, in update_class\n",
      "    update_instances(old, new)\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 319, in update_instances\n",
      "    object.__setattr__(ref, \"__class__\", new)\n",
      "TypeError: __class__ assignment: 'uname_result' object layout differs from 'uname_result'\n",
      "]\n",
      "[autoreload of ctypes._endian failed: Traceback (most recent call last):\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 273, in check\n",
      "    superreload(m, reload, self.old_objects)\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 496, in superreload\n",
      "    update_generic(old_obj, new_obj)\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 393, in update_generic\n",
      "    update(a, b)\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 361, in update_class\n",
      "    update_instances(old, new)\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 319, in update_instances\n",
      "    object.__setattr__(ref, \"__class__\", new)\n",
      "TypeError: can't apply this __setattr__ to _swapped_meta object\n",
      "]\n",
      "[autoreload of ctypes failed: Traceback (most recent call last):\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 273, in check\n",
      "    superreload(m, reload, self.old_objects)\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 496, in superreload\n",
      "    update_generic(old_obj, new_obj)\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 393, in update_generic\n",
      "    update(a, b)\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 345, in update_class\n",
      "    if update_generic(old_obj, new_obj):\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 393, in update_generic\n",
      "    update(a, b)\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 345, in update_class\n",
      "    if update_generic(old_obj, new_obj):\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 393, in update_generic\n",
      "    update(a, b)\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 345, in update_class\n",
      "    if update_generic(old_obj, new_obj):\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 393, in update_generic\n",
      "    update(a, b)\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 345, in update_class\n",
      "    if update_generic(old_obj, new_obj):\n",
      "RecursionError: maximum recursion depth exceeded in comparison\n",
      "]\n",
      "[autoreload of socket failed: Traceback (most recent call last):\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 273, in check\n",
      "    superreload(m, reload, self.old_objects)\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 496, in superreload\n",
      "    update_generic(old_obj, new_obj)\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 393, in update_generic\n",
      "    update(a, b)\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 361, in update_class\n",
      "    update_instances(old, new)\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 319, in update_instances\n",
      "    object.__setattr__(ref, \"__class__\", new)\n",
      "TypeError: __class__ assignment: 'socket' object layout differs from 'socket'\n",
      "]\n",
      "[autoreload of ssl failed: Traceback (most recent call last):\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 273, in check\n",
      "    superreload(m, reload, self.old_objects)\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 496, in superreload\n",
      "    update_generic(old_obj, new_obj)\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 393, in update_generic\n",
      "    update(a, b)\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 361, in update_class\n",
      "    update_instances(old, new)\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 319, in update_instances\n",
      "    object.__setattr__(ref, \"__class__\", new)\n",
      "TypeError: __class__ assignment: '_SSLMethod' object layout differs from '_SSLMethod'\n",
      "]\n",
      "[autoreload of ast failed: Traceback (most recent call last):\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 273, in check\n",
      "    superreload(m, reload, self.old_objects)\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 496, in superreload\n",
      "    update_generic(old_obj, new_obj)\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 393, in update_generic\n",
      "    update(a, b)\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 361, in update_class\n",
      "    update_instances(old, new)\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 319, in update_instances\n",
      "    object.__setattr__(ref, \"__class__\", new)\n",
      "TypeError: can't apply this __setattr__ to _ABC object\n",
      "]\n",
      "[autoreload of inspect failed: Traceback (most recent call last):\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 273, in check\n",
      "    superreload(m, reload, self.old_objects)\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 496, in superreload\n",
      "    update_generic(old_obj, new_obj)\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 393, in update_generic\n",
      "    update(a, b)\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 361, in update_class\n",
      "    update_instances(old, new)\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 319, in update_instances\n",
      "    object.__setattr__(ref, \"__class__\", new)\n",
      "TypeError: __class__ assignment: '_ParameterKind' object layout differs from '_ParameterKind'\n",
      "]\n",
      "[autoreload of asyncio.coroutines failed: Traceback (most recent call last):\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 273, in check\n",
      "    superreload(m, reload, self.old_objects)\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 471, in superreload\n",
      "    module = reload(module)\n",
      "  File \"/opt/conda/lib/python3.9/importlib/__init__.py\", line 169, in reload\n",
      "    _bootstrap._exec(spec, module)\n",
      "  File \"<frozen importlib._bootstrap>\", line 613, in _exec\n",
      "  File \"<frozen importlib._bootstrap_external>\", line 850, in exec_module\n",
      "  File \"<frozen importlib._bootstrap>\", line 228, in _call_with_frames_removed\n",
      "  File \"/opt/conda/lib/python3.9/asyncio/coroutines.py\", line 173, in <module>\n",
      "    collections.abc.Coroutine, CoroWrapper)\n",
      "  File \"/opt/conda/lib/python3.9/collections/__init__.py\", line 68, in __getattr__\n",
      "    raise AttributeError(f'module {__name__!r} has no attribute {name!r}')\n",
      "AttributeError: module 'collections' has no attribute 'abc'\n",
      "]\n",
      "[autoreload of asyncio.tasks failed: Traceback (most recent call last):\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 273, in check\n",
      "    superreload(m, reload, self.old_objects)\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 471, in superreload\n",
      "    module = reload(module)\n",
      "  File \"/opt/conda/lib/python3.9/importlib/__init__.py\", line 169, in reload\n",
      "    _bootstrap._exec(spec, module)\n",
      "  File \"<frozen importlib._bootstrap>\", line 613, in _exec\n",
      "  File \"<frozen importlib._bootstrap_external>\", line 850, in exec_module\n",
      "  File \"<frozen importlib._bootstrap>\", line 228, in _call_with_frames_removed\n",
      "  File \"/opt/conda/lib/python3.9/asyncio/tasks.py\", line 368, in <module>\n",
      "    FIRST_COMPLETED = concurrent.futures.FIRST_COMPLETED\n",
      "AttributeError: module 'concurrent' has no attribute 'futures'\n",
      "]\n",
      "[autoreload of asyncio failed: Traceback (most recent call last):\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 273, in check\n",
      "    superreload(m, reload, self.old_objects)\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 471, in superreload\n",
      "    module = reload(module)\n",
      "  File \"/opt/conda/lib/python3.9/importlib/__init__.py\", line 169, in reload\n",
      "    _bootstrap._exec(spec, module)\n",
      "  File \"<frozen importlib._bootstrap>\", line 613, in _exec\n",
      "  File \"<frozen importlib._bootstrap_external>\", line 850, in exec_module\n",
      "  File \"<frozen importlib._bootstrap>\", line 228, in _call_with_frames_removed\n",
      "  File \"/opt/conda/lib/python3.9/asyncio/__init__.py\", line 27, in <module>\n",
      "    __all__ = (base_events.__all__ +\n",
      "NameError: name 'base_events' is not defined\n",
      "]\n",
      "[autoreload of urllib.parse failed: Traceback (most recent call last):\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 273, in check\n",
      "    superreload(m, reload, self.old_objects)\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 496, in superreload\n",
      "    update_generic(old_obj, new_obj)\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 393, in update_generic\n",
      "    update(a, b)\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 345, in update_class\n",
      "    if update_generic(old_obj, new_obj):\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 393, in update_generic\n",
      "    update(a, b)\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 345, in update_class\n",
      "    if update_generic(old_obj, new_obj):\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 393, in update_generic\n",
      "    update(a, b)\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 345, in update_class\n",
      "    if update_generic(old_obj, new_obj):\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 393, in update_generic\n",
      "    update(a, b)\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 345, in update_class\n",
      "    if update_generic(old_obj, new_obj):\n",
      "RecursionError: maximum recursion depth exceeded\n",
      "]\n",
      "[autoreload of pathlib failed: Traceback (most recent call last):\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 273, in check\n",
      "    superreload(m, reload, self.old_objects)\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 496, in superreload\n",
      "    update_generic(old_obj, new_obj)\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 393, in update_generic\n",
      "    update(a, b)\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 361, in update_class\n",
      "    update_instances(old, new)\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 319, in update_instances\n",
      "    object.__setattr__(ref, \"__class__\", new)\n",
      "TypeError: __class__ assignment: 'PosixPath' object layout differs from 'PosixPath'\n",
      "]\n",
      "[autoreload of uuid failed: Traceback (most recent call last):\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 273, in check\n",
      "    superreload(m, reload, self.old_objects)\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 496, in superreload\n",
      "    update_generic(old_obj, new_obj)\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 393, in update_generic\n",
      "    update(a, b)\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 361, in update_class\n",
      "    update_instances(old, new)\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 319, in update_instances\n",
      "    object.__setattr__(ref, \"__class__\", new)\n",
      "TypeError: __class__ assignment: 'UUID' object layout differs from 'UUID'\n",
      "]\n",
      "[autoreload of email.header failed: Traceback (most recent call last):\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 273, in check\n",
      "    superreload(m, reload, self.old_objects)\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 471, in superreload\n",
      "    module = reload(module)\n",
      "  File \"/opt/conda/lib/python3.9/importlib/__init__.py\", line 169, in reload\n",
      "    _bootstrap._exec(spec, module)\n",
      "  File \"<frozen importlib._bootstrap>\", line 613, in _exec\n",
      "  File \"<frozen importlib._bootstrap_external>\", line 850, in exec_module\n",
      "  File \"<frozen importlib._bootstrap>\", line 228, in _call_with_frames_removed\n",
      "  File \"/opt/conda/lib/python3.9/email/header.py\", line 57, in <module>\n",
      "    _max_append = email.quoprimime._max_append\n",
      "AttributeError: module 'email' has no attribute 'quoprimime'\n",
      "]\n",
      "[autoreload of sqlite3.dbapi2 failed: Traceback (most recent call last):\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 273, in check\n",
      "    superreload(m, reload, self.old_objects)\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 471, in superreload\n",
      "    module = reload(module)\n",
      "  File \"/opt/conda/lib/python3.9/importlib/__init__.py\", line 169, in reload\n",
      "    _bootstrap._exec(spec, module)\n",
      "  File \"<frozen importlib._bootstrap>\", line 613, in _exec\n",
      "  File \"<frozen importlib._bootstrap_external>\", line 850, in exec_module\n",
      "  File \"<frozen importlib._bootstrap>\", line 228, in _call_with_frames_removed\n",
      "  File \"/opt/conda/lib/python3.9/sqlite3/dbapi2.py\", line 54, in <module>\n",
      "    collections.abc.Sequence.register(Row)\n",
      "  File \"/opt/conda/lib/python3.9/collections/__init__.py\", line 68, in __getattr__\n",
      "    raise AttributeError(f'module {__name__!r} has no attribute {name!r}')\n",
      "AttributeError: module 'collections' has no attribute 'abc'\n",
      "]\n",
      "[autoreload of xml.dom.minidom failed: Traceback (most recent call last):\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 273, in check\n",
      "    superreload(m, reload, self.old_objects)\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 471, in superreload\n",
      "    module = reload(module)\n",
      "  File \"/opt/conda/lib/python3.9/importlib/__init__.py\", line 169, in reload\n",
      "    _bootstrap._exec(spec, module)\n",
      "  File \"<frozen importlib._bootstrap>\", line 613, in _exec\n",
      "  File \"<frozen importlib._bootstrap_external>\", line 850, in exec_module\n",
      "  File \"<frozen importlib._bootstrap>\", line 228, in _call_with_frames_removed\n",
      "  File \"/opt/conda/lib/python3.9/xml/dom/minidom.py\", line 30, in <module>\n",
      "    _nodeTypes_with_children = (xml.dom.Node.ELEMENT_NODE,\n",
      "AttributeError: module 'xml' has no attribute 'dom'\n",
      "]\n",
      "[autoreload of http failed: Traceback (most recent call last):\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 273, in check\n",
      "    superreload(m, reload, self.old_objects)\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 496, in superreload\n",
      "    update_generic(old_obj, new_obj)\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 393, in update_generic\n",
      "    update(a, b)\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 361, in update_class\n",
      "    update_instances(old, new)\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 319, in update_instances\n",
      "    object.__setattr__(ref, \"__class__\", new)\n",
      "TypeError: __class__ assignment: 'HTTPStatus' object layout differs from 'HTTPStatus'\n",
      "]\n",
      "[autoreload of http.client failed: Traceback (most recent call last):\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 273, in check\n",
      "    superreload(m, reload, self.old_objects)\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 471, in superreload\n",
      "    module = reload(module)\n",
      "  File \"/opt/conda/lib/python3.9/importlib/__init__.py\", line 169, in reload\n",
      "    _bootstrap._exec(spec, module)\n",
      "  File \"<frozen importlib._bootstrap>\", line 613, in _exec\n",
      "  File \"<frozen importlib._bootstrap_external>\", line 850, in exec_module\n",
      "  File \"<frozen importlib._bootstrap>\", line 228, in _call_with_frames_removed\n",
      "  File \"/opt/conda/lib/python3.9/http/client.py\", line 178, in <module>\n",
      "    class HTTPMessage(email.message.Message):\n",
      "AttributeError: module 'email' has no attribute 'message'\n",
      "]\n",
      "[autoreload of urllib.error failed: Traceback (most recent call last):\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 273, in check\n",
      "    superreload(m, reload, self.old_objects)\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 471, in superreload\n",
      "    module = reload(module)\n",
      "  File \"/opt/conda/lib/python3.9/importlib/__init__.py\", line 169, in reload\n",
      "    _bootstrap._exec(spec, module)\n",
      "  File \"<frozen importlib._bootstrap>\", line 613, in _exec\n",
      "  File \"<frozen importlib._bootstrap_external>\", line 850, in exec_module\n",
      "  File \"<frozen importlib._bootstrap>\", line 228, in _call_with_frames_removed\n",
      "  File \"/opt/conda/lib/python3.9/urllib/error.py\", line 35, in <module>\n",
      "    class HTTPError(URLError, urllib.response.addinfourl):\n",
      "AttributeError: module 'urllib' has no attribute 'response'\n",
      "]\n",
      "[autoreload of urllib.request failed: Traceback (most recent call last):\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 273, in check\n",
      "    superreload(m, reload, self.old_objects)\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 471, in superreload\n",
      "    module = reload(module)\n",
      "  File \"/opt/conda/lib/python3.9/importlib/__init__.py\", line 169, in reload\n",
      "    _bootstrap._exec(spec, module)\n",
      "  File \"<frozen importlib._bootstrap>\", line 613, in _exec\n",
      "  File \"<frozen importlib._bootstrap_external>\", line 850, in exec_module\n",
      "  File \"<frozen importlib._bootstrap>\", line 228, in _call_with_frames_removed\n",
      "  File \"/opt/conda/lib/python3.9/urllib/request.py\", line 1379, in <module>\n",
      "    if hasattr(http.client, 'HTTPSConnection'):\n",
      "AttributeError: module 'http' has no attribute 'client'\n",
      "]\n",
      "[autoreload of http.server failed: Traceback (most recent call last):\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 273, in check\n",
      "    superreload(m, reload, self.old_objects)\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 471, in superreload\n",
      "    module = reload(module)\n",
      "  File \"/opt/conda/lib/python3.9/importlib/__init__.py\", line 169, in reload\n",
      "    _bootstrap._exec(spec, module)\n",
      "  File \"<frozen importlib._bootstrap>\", line 613, in _exec\n",
      "  File \"<frozen importlib._bootstrap_external>\", line 850, in exec_module\n",
      "  File \"<frozen importlib._bootstrap>\", line 228, in _call_with_frames_removed\n",
      "  File \"/opt/conda/lib/python3.9/http/server.py\", line 147, in <module>\n",
      "    class BaseHTTPRequestHandler(socketserver.StreamRequestHandler):\n",
      "  File \"/opt/conda/lib/python3.9/http/server.py\", line 634, in BaseHTTPRequestHandler\n",
      "    MessageClass = http.client.HTTPMessage\n",
      "AttributeError: module 'http' has no attribute 'client'\n",
      "]\n",
      "[autoreload of multiprocessing.process failed: Traceback (most recent call last):\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 273, in check\n",
      "    superreload(m, reload, self.old_objects)\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 496, in superreload\n",
      "    update_generic(old_obj, new_obj)\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 393, in update_generic\n",
      "    update(a, b)\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 361, in update_class\n",
      "    update_instances(old, new)\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/extensions/autoreload.py\", line 319, in update_instances\n",
      "    object.__setattr__(ref, \"__class__\", new)\n",
      "TypeError: __class__ assignment: 'AuthenticationString' object layout differs from 'AuthenticationString'\n",
      "]\n"
     ]
    },
    {
     "ename": "TypeError",
     "evalue": "descriptor '_str' for 'PurePath' objects doesn't apply to a 'PosixPath' object",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[1;32m/root/capsule/code/dev_eye_data.ipynb Cell 7\u001b[0m line \u001b[0;36m1\n\u001b[0;32m----> <a href='vscode-notebook-cell://codeocean.allenneuraldynamics.org/root/capsule/code/dev_eye_data.ipynb#X10sdnNjb2RlLXJlbW90ZQ%3D%3D?line=0'>1</a>\u001b[0m eye_tracking_data \u001b[39m=\u001b[39m load_eye_tracking_hdf(eye_h5_fn)\n",
      "\u001b[1;32m/root/capsule/code/dev_eye_data.ipynb Cell 7\u001b[0m line \u001b[0;36m2\n\u001b[1;32m     <a href='vscode-notebook-cell://codeocean.allenneuraldynamics.org/root/capsule/code/dev_eye_data.ipynb#X10sdnNjb2RlLXJlbW90ZQ%3D%3D?line=25'>26</a>\u001b[0m eye_tracking_dfs \u001b[39m=\u001b[39m []\n\u001b[1;32m     <a href='vscode-notebook-cell://codeocean.allenneuraldynamics.org/root/capsule/code/dev_eye_data.ipynb#X10sdnNjb2RlLXJlbW90ZQ%3D%3D?line=26'>27</a>\u001b[0m \u001b[39mfor\u001b[39;00m field_name \u001b[39min\u001b[39;00m eye_tracking_fields:\n\u001b[0;32m---> <a href='vscode-notebook-cell://codeocean.allenneuraldynamics.org/root/capsule/code/dev_eye_data.ipynb#X10sdnNjb2RlLXJlbW90ZQ%3D%3D?line=27'>28</a>\u001b[0m     field_data \u001b[39m=\u001b[39m pd\u001b[39m.\u001b[39;49mread_hdf(eye_tracking_file, key\u001b[39m=\u001b[39;49mfield_name)\n\u001b[1;32m     <a href='vscode-notebook-cell://codeocean.allenneuraldynamics.org/root/capsule/code/dev_eye_data.ipynb#X10sdnNjb2RlLXJlbW90ZQ%3D%3D?line=28'>29</a>\u001b[0m     new_col_name_map \u001b[39m=\u001b[39m {col_name: \u001b[39mf\u001b[39m\u001b[39m\"\u001b[39m\u001b[39m{\u001b[39;00mfield_name\u001b[39m}\u001b[39;00m\u001b[39m_\u001b[39m\u001b[39m{\u001b[39;00mcol_name\u001b[39m}\u001b[39;00m\u001b[39m\"\u001b[39m\n\u001b[1;32m     <a href='vscode-notebook-cell://codeocean.allenneuraldynamics.org/root/capsule/code/dev_eye_data.ipynb#X10sdnNjb2RlLXJlbW90ZQ%3D%3D?line=29'>30</a>\u001b[0m                         \u001b[39mfor\u001b[39;00m col_name \u001b[39min\u001b[39;00m field_data\u001b[39m.\u001b[39mcolumns}\n\u001b[1;32m     <a href='vscode-notebook-cell://codeocean.allenneuraldynamics.org/root/capsule/code/dev_eye_data.ipynb#X10sdnNjb2RlLXJlbW90ZQ%3D%3D?line=30'>31</a>\u001b[0m     field_data\u001b[39m.\u001b[39mrename(new_col_name_map, axis\u001b[39m=\u001b[39m\u001b[39m1\u001b[39m, inplace\u001b[39m=\u001b[39m\u001b[39mTrue\u001b[39;00m)\n",
      "File \u001b[0;32m/opt/conda/lib/python3.9/site-packages/pandas/io/pytables.py:411\u001b[0m, in \u001b[0;36mread_hdf\u001b[0;34m(path_or_buf, key, mode, errors, where, start, stop, columns, iterator, chunksize, **kwargs)\u001b[0m\n\u001b[1;32m    409\u001b[0m     auto_close \u001b[39m=\u001b[39m \u001b[39mFalse\u001b[39;00m\n\u001b[1;32m    410\u001b[0m \u001b[39melse\u001b[39;00m:\n\u001b[0;32m--> 411\u001b[0m     path_or_buf \u001b[39m=\u001b[39m stringify_path(path_or_buf)\n\u001b[1;32m    412\u001b[0m     \u001b[39mif\u001b[39;00m \u001b[39mnot\u001b[39;00m \u001b[39misinstance\u001b[39m(path_or_buf, \u001b[39mstr\u001b[39m):\n\u001b[1;32m    413\u001b[0m         \u001b[39mraise\u001b[39;00m \u001b[39mNotImplementedError\u001b[39;00m(\n\u001b[1;32m    414\u001b[0m             \u001b[39m\"\u001b[39m\u001b[39mSupport for generic buffers has not been implemented.\u001b[39m\u001b[39m\"\u001b[39m\n\u001b[1;32m    415\u001b[0m         )\n",
      "File \u001b[0;32m/opt/conda/lib/python3.9/site-packages/pandas/io/common.py:278\u001b[0m, in \u001b[0;36mstringify_path\u001b[0;34m(filepath_or_buffer, convert_file_like)\u001b[0m\n\u001b[1;32m    275\u001b[0m     \u001b[39mreturn\u001b[39;00m cast(BaseBufferT, filepath_or_buffer)\n\u001b[1;32m    277\u001b[0m \u001b[39mif\u001b[39;00m \u001b[39misinstance\u001b[39m(filepath_or_buffer, os\u001b[39m.\u001b[39mPathLike):\n\u001b[0;32m--> 278\u001b[0m     filepath_or_buffer \u001b[39m=\u001b[39m filepath_or_buffer\u001b[39m.\u001b[39;49m__fspath__()\n\u001b[1;32m    279\u001b[0m \u001b[39mreturn\u001b[39;00m _expand_user(filepath_or_buffer)\n",
      "File \u001b[0;32m/opt/conda/lib/python3.9/pathlib.py:753\u001b[0m, in \u001b[0;36mPurePath.__fspath__\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    752\u001b[0m \u001b[39mdef\u001b[39;00m \u001b[39m__fspath__\u001b[39m(\u001b[39mself\u001b[39m):\n\u001b[0;32m--> 753\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mstr\u001b[39;49m(\u001b[39mself\u001b[39;49m)\n",
      "File \u001b[0;32m/opt/conda/lib/python3.9/pathlib.py:746\u001b[0m, in \u001b[0;36mPurePath.__str__\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    743\u001b[0m \u001b[39m\u001b[39m\u001b[39m\"\"\"Return the string representation of the path, suitable for\u001b[39;00m\n\u001b[1;32m    744\u001b[0m \u001b[39mpassing to system calls.\"\"\"\u001b[39;00m\n\u001b[1;32m    745\u001b[0m \u001b[39mtry\u001b[39;00m:\n\u001b[0;32m--> 746\u001b[0m     \u001b[39mreturn\u001b[39;00m \u001b[39mself\u001b[39;49m\u001b[39m.\u001b[39;49m_str\n\u001b[1;32m    747\u001b[0m \u001b[39mexcept\u001b[39;00m \u001b[39mAttributeError\u001b[39;00m:\n\u001b[1;32m    748\u001b[0m     \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_str \u001b[39m=\u001b[39m \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_format_parsed_parts(\u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_drv, \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_root,\n\u001b[1;32m    749\u001b[0m                                           \u001b[39mself\u001b[39m\u001b[39m.\u001b[39m_parts) \u001b[39mor\u001b[39;00m \u001b[39m'\u001b[39m\u001b[39m.\u001b[39m\u001b[39m'\u001b[39m\n",
      "\u001b[0;31mTypeError\u001b[0m: descriptor '_str' for 'PurePath' objects doesn't apply to a 'PosixPath' object"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Traceback (most recent call last):\n",
      "  File \"/opt/conda/lib/python3.9/runpy.py\", line 197, in _run_module_as_main\n",
      "    return _run_code(code, main_globals, None,\n",
      "  File \"/opt/conda/lib/python3.9/runpy.py\", line 87, in _run_code\n",
      "    exec(code, run_globals)\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/ipykernel_launcher.py\", line 17, in <module>\n",
      "    app.launch_new_instance()\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/traitlets/config/application.py\", line 1043, in launch_instance\n",
      "    app.start()\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/ipykernel/kernelapp.py\", line 725, in start\n",
      "    self.io_loop.start()\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/tornado/platform/asyncio.py\", line 215, in start\n",
      "    self.asyncio_loop.run_forever()\n",
      "  File \"/opt/conda/lib/python3.9/asyncio/base_events.py\", line 601, in run_forever\n",
      "    self._run_once()\n",
      "  File \"/opt/conda/lib/python3.9/asyncio/base_events.py\", line 1891, in _run_once\n",
      "    if handle._cancelled:\n",
      "TypeError: descriptor '_cancelled' for 'Handle' objects doesn't apply to a 'Handle' object\n",
      "Error in atexit._run_exitfuncs:\n",
      "Traceback (most recent call last):\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/core/interactiveshell.py\", line 3827, in atexit_operations\n",
      "    self._atexit_once()\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/core/interactiveshell.py\", line 3806, in _atexit_once\n",
      "    self.reset(new_session=False)\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/core/interactiveshell.py\", line 1371, in reset\n",
      "    self.history_manager.reset(new_session)\n",
      "  File \"/opt/conda/lib/python3.9/site-packages/IPython/core/history.py\", line 594, in reset\n",
      "    self.dir_hist[:] = [Path.cwd()]\n",
      "  File \"/opt/conda/lib/python3.9/pathlib.py\", line 1136, in cwd\n",
      "    return cls(os.getcwd())\n",
      "  File \"/opt/conda/lib/python3.9/pathlib.py\", line 1082, in __new__\n",
      "    self = cls._from_parts(args, init=False)\n",
      "  File \"/opt/conda/lib/python3.9/pathlib.py\", line 707, in _from_parts\n",
      "    drv, root, parts = self._parse_args(args)\n",
      "  File \"/opt/conda/lib/python3.9/pathlib.py\", line 700, in _parse_args\n",
      "    return cls._flavour.parse_parts(parts)\n",
      "AttributeError: type object 'Path' has no attribute '_flavour'\n"
     ]
    },
    {
     "ename": "",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31mCannot execute code, session has been disposed. Please try restarting the Kernel."
     ]
    },
    {
     "ename": "",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31mThe Kernel crashed while executing code in the the current cell or a previous cell. Please review the code in the cell(s) to identify a possible cause of the failure. Click <a href='https://aka.ms/vscodeJupyterKernelCrash'>here</a> for more info. View Jupyter <a href='command:jupyter.viewOutput'>log</a> for further details."
     ]
    }
   ],
   "source": [
    "eye_tracking_data = load_eye_tracking_hdf(eye_h5_fn)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
